File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head: return None
        stack = []
        while head:
            stack.append(head)
            head = head.next
        head = stack.pop()
        res = head
        i = len(stack)-1
        while i>=0:
            node =  stack.pop()
            head.next = node
            i-=1
            head = head.next
        head.next =None
        return res

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if(lists.length == 0){
            return null;
        }
        int l = 0, r = lists.length -1;
        return divide(lists, l, r);
    }

    private ListNode divide(ListNode[] list, int left, int right){
        int mid = (left +right)/2;
        if (left > right || right < left){
            return null;
        }
        if (left == right){
            return list[left];
        }
        ListNode leftNode = divide(list, left, mid);
        ListNode rightNode = divide(list, mid +1, right);

        return combine(leftNode, rightNode);
    }

    private ListNode combine(ListNode left, ListNode right){
        ListNode merged = new ListNode(1);
        var temp = merged;
        while (left != null && right != null){
            if (left.val < right.val){
                merged.next = new ListNode(left.val);
                left = left.next;
            }
            else{
                merged.next = new ListNode(right.val);
                right = right.next;
            }
            merged = merged.next;
        }
        if(left != null){
            merged.next  = left;
        }
        else if(right != null){
            merged.next = right;
        }
        return temp.next;
    }
}
